policy/pa.go

```diff
@@ -388,27 +402,60 @@ func (pa *AuthorityImpl) willingToIssue(id identifier.ACMEIdentifier) error {
 //     blocklist entry for "foo.example.com" should prevent issuance for
 //     "*.example.com")
 //
-// If any of the identifiers are not valid then an error with suberrors specific
-// to the rejected identifiers will be returned.
-func (pa *AuthorityImpl) WillingToIssueWildcards(idents []identifier.ACMEIdentifier) error {
+// If any of the domains are not valid then an error with suberrors specific
+// to the rejected domains will be returned.
+func (pa *AuthorityImpl) WillingToIssue(domains []string) error {
 	var subErrors []berrors.SubBoulderError
-	for _, ident := range idents {
-		err := pa.willingToIssueWildcard(ident)
-		if err != nil {
-			var bErr *berrors.BoulderError
-			if errors.As(err, &bErr) {
-				subErrors = append(subErrors, berrors.SubBoulderError{
-					Identifier:   ident,
-					BoulderError: bErr})
-			} else {
-				subErrors = append(subErrors, berrors.SubBoulderError{
-					Identifier: ident,
-					BoulderError: &berrors.BoulderError{
-						Type:   berrors.RejectedIdentifier,
-						Detail: err.Error(),
-					}})
+	appendSubError := func(name string, err error) {
+		var bErr *berrors.BoulderError
+		if errors.As(err, &bErr) {
+			subErrors = append(subErrors, berrors.SubBoulderError{
+				Identifier:   identifier.DNSIdentifier(name),
+				BoulderError: bErr,
+			})
+		} else {
+			subErrors = append(subErrors, berrors.SubBoulderError{
+				Identifier: identifier.DNSIdentifier(name),
+				BoulderError: &berrors.BoulderError{
+					Type:   berrors.RejectedIdentifier,
+					Detail: err.Error(),
+				},
+			})
+		}
+	}
+
+	for _, domain := range domains {
+		if strings.Count(domain, "*") > 0 {
+			// Domain contains a wildcard, check that it is valid.
+			err := ValidWildcardDomain(domain)
+			if err != nil {
+				appendSubError(domain, err)
+				continue
+			}
+
+			// The base domain is the wildcard request with the `*.` prefix removed
+			baseDomain := strings.TrimPrefix(domain, "*.")
+
+			// The base domain can't be in the wildcard exact blocklist
+			err = pa.checkWildcardHostList(baseDomain)
+			if err != nil {
+				appendSubError(domain, err)
+				continue
+			}
+		} else {
+			// Validate that the domain is well-formed.
+			err := ValidDomain(domain)
+			if err != nil {
+				appendSubError(domain, err)
+				continue
 			}
 		}
+		// Require no exact match against hostname block lists
```

Comment:
```suggestion
		// Require no match against hostname block lists
```

It looks like this PR introduces the word "exact" here which is incorrect. `checkHostLists` checks for both subdomain matches and exact matches in the blocklists.